src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -39,10 +43,79 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, ImmutableArray<IDiagno
             // name binding is done
             // allow type queries now
             paramsSymbolContext.Unlock();
+            this.AssignmentsByDeclaration = InitializeDeclarationToAssignmentDictionary();
+            this.DeclarationsByAssignment = InitializeAssignmentToDeclarationDictionary();
             // lazy load single use diagnostic set
             this.AllDiagnostics = new Lazy<ImmutableArray<IDiagnostic>>(() => GetAllDiagnostics());
         }
 
+        public ImmutableArray<IDiagnostic> GetAllDiagnostics()
+            => BicepParamFile.ProgramSyntax.GetParseDiagnostics()
+                .Concat(ParamsTypeManager.GetAllDiagnostics())
+                .Concat(this.CompilationLoadDiagnostics)
+                .Concat(GetAdditionalSemanticDiagnostics()).ToImmutableArray();
+            
+        /// <summary>
+        /// Gets the file that was compiled.
+        /// </summary>
+        public ParamFileSymbol Root => this.ParamBinder.ParamFileSymbol;
+
+        public ParameterSymbol? GetParameterDeclaration(ParameterAssignmentSymbol parameterAssignmentSymbol)
+        {
+            return this.DeclarationsByAssignment[parameterAssignmentSymbol];
+        }
+
+        public ParameterAssignmentSymbol? GetParameterAssignment(ParameterSymbol parameterSymbol)
+        {
+            if (this.AssignmentsByDeclaration is null)
+            {
+                return null;
+            }
+            return this.AssignmentsByDeclaration[parameterSymbol];
+        }
+
+        private ImmutableDictionary<ParameterSymbol, ParameterAssignmentSymbol?>? InitializeDeclarationToAssignmentDictionary()
+        {
+            if (this.BicepCompilation is null)
+            {
+                return null;
```

Comment:
We can make assignmentsByDeclaration a non-null dictionary and return an empty dictionary. This avoids null checks on the dictionary itself in other places.